### ISSUE TITLE
tag-fields (or lookup-fields / relationship-fields)

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#1324]
Introduce a **generic** way to declare fields to be "lookup-fields" that are treated and shown as if tags, thus "semantic tags" to model lookup relationships from one tiddler pointing to another in a qualified manner.

**the problem**

One may wish to assign an *author* to a tiddler. However, assigning the *author* only via either of the following methods has pitfalls...

1. a tag named `<AuthorName>`
    * **problem:** when aggregating lists, there is no (easy way) to programatically identify...
        * "Is there and what is the author associated with this tiddler?"
1. a field defined as `author:<AuthorName>`
    * **problem:** the *author* information is buried, hence not as easily accessible, visible, listed
        * also, with no tag-popup, typos are more frequent

**the solution**

Therefore, it would be highly practical if one could define a set of fields to act as if tags, in a way "qualified tags"... and are thus shown in a tiddler's taglist with a clickable button as described below which, however, actually looks up on the corresponding tag-field rather than an actual tag.

**a tag-field-button-popup**

If one were to click such a tag-field-button in a tiddler's taglist, it would show...

* a link to a tiddler called `<tag-field>`, or rather an associated tiddler title
    * the title for the tiddler corresponding to the *author* field could be **Authors** instead of the user being forced to use a lowercase **author**
    * this tiddler could then be used to output a list of all different *author* values and perhaps associated tiddlers, similar to how the timeline works, only just not based on dates
* a link to a tiddler corresponding to `<TagFieldValue>`, e.g. the `<AuthorName>`
* a list with all tiddlers where `[<tag-field>[<TagFieldValue>]]`
    * e.g. where the value of *author* is the same as the current value, namely `<AuthorName>`
* a list of any other, not yet listed tiddlers where `[tag[<TagFieldValue>]]`, e.g. tagging to `<AuthorName>`

This could even be a qualified button displayed with a title of `author:<AuthorName>`.

**"multiple inheritance"**

Just like regular tags, in the sense of a tag modeling a **parent => child** relationship, it may be desireable to be able to assign more than one tiddler as a value to a tag-field. This would have implications to the popup above as it would raise the question what entries the popup should list, e.g. tiddlers for each *author* or tiddlers with the same combination of the current *authors*, probably the former. Thus, a number of the above qualified tag-field-buttons could be rendered...

*tags*: `tagFoo` `author:<AuthorName1>` `author:<AuthorName2>`

**independent plugin**

If this was not desireable as a core functionality, which I think it is in so many ways, one would have to be able to update and add to a dom element, e.g. to programatically add tag-field-buttons to the default taglist. How would and can that work? It seems to me there needs to be a way for one widget to hook into another widget as a "dependent widget" and thus tell that other widget to also update the dependent widget when being updated. Of course, a dependent widget must not (be allowed to) update the calling widget, otherwise we'd be endlessly running in circles.

**Comment by Jermolene** [comment]
Hi @tobibeer interesting.

> problem: the author information is buried, hence not as easily accessible, visible, listed also, with no tag-popup, typos are more frequent

This proposal seems to be predicated on the difficulty of these problems but I don't think they are very hard to overcome. It's true that fields aren't particularly visible with the default templates, but those templates can be changed.

> If this was not desireable as a core functionality, which I think it is in so many ways

It's a bit early to tell, given that it's just a proposal at the moment. The way to explore this is via a plugin.

**Comment by twMat** [comment]
@Jermolene 

I'm curious (and please tell me if this is better as a group discussion) - siimilar proposals, albeit never as specified and well expressed as here, have been put up previously and I recall you've expressed interest already then. Can I ask what your general reflections are on this idea, i.e the idea with "semantic tags", "tag types", "using fields as tags" and related? To me (and obviously Tobias, and a few others) it seems like a *very* powerful tool and a valuable addition to TW, but somehow I sense it never really... I dunno... "catches on"? I hope it's ok to ask this as I must assume you have some insights I've just not understood yet? Or am I misunderstanding something on the matter? 
(better for google groups?)

Thank you Jeremy!

**Comment by tobibeer** [comment]
I think it's both a matter of effort and imagining use-cases where this will be great.

Besides the above mentioned enhanced "qualified tag pill", one major usecase would be a representation in TiddlyMap, e.g. showing relations such as "author of" and "authored by"... on top of any generic relations like "tagged by", "tagging to", "linked by", "linking to", "listed by", "listing", etc...

**Comment by Jermolene** [comment]
> I'm curious (and please tell me if this is better as a group discussion) - siimilar proposals, albeit never as specified and well expressed as here, have been put up previously and I recall you've expressed interest already then.

I'm not uninterested in this proposal, I think it's an interesting area for exploration. I think that the best way to explore it is via a plugin: I think a lot of what @tobibeer is suggesting can be readily explored with custom segments for the view and edit templates. Then when we've got some experience of how well it works we can consider whether and how to fold it into the core.

> Can I ask what your general reflections are on this idea, i.e the idea with "semantic tags", "tag types", "using fields as tags" and related? To me (and obviously Tobias, and a few others) it seems like a very powerful tool and a valuable addition to TW, but somehow I sense it never really... I dunno... "catches on"? I hope it's ok to ask this as I must assume you have some insights I've just not understood yet? Or am I misunderstanding something on the matter? 

I'd like it to be easy to map RDF-like structures onto TiddlyWiki. I think the primitives required are all already in the core, but there's not necessarily a suitable user interface.

**Comment by twMat** [comment]
Interestingly, at least for me, as I drew the illustrations below it became obvious to me how much I agree with that a plugin probably is the best way to go about it. 

@tobibeer - I tried to make an illustration from your description. Please tell me what to change in it.
Note: Regarding *"the title for the tiddler corresponding to the author field could be Authors instead of the user being forced to use a lowercase author"* - I have expressed this as [[Authors|author]] even if I understand that's not really correct.

![tagfields_tobias](https://cloud.githubusercontent.com/assets/4534056/5689692/eab22e14-986a-11e4-8476-a5d9b01c4dda.png)

And here follows my own suggestion on the matter. It looks much simpler but this is mainly due to omitting the ´generic code´ parameters. (And the font-size difference is due to github zooming). Note that the right-lower column lists differently than in your version (I mention this because I think it would be easy to miss otherwise).

![tagfields_simple](https://cloud.githubusercontent.com/assets/4534056/5689711/2a7bd8b0-986b-11e4-8a87-0cab8f3d15a5.png)

A plugin does feel justified as I see how many variable parts there are in this.
...

BTW, as I briefly mentioned in [this](https://groups.google.com/d/msg/tiddlywiki/sCTtIB5ic9c/ZMVYGZcp1X8J) discussion, what would happen if regular tags were skipped alltogether and we only used these tag-fields instead *but* with a special default fieldname called "tag" that, unless changed into something else (like "author") simply stays invisible effectively making it all look like a tags currently do? I.e *<fieldname>:<fieldvalue>* becomes *tag:<fieldvalue>* as default.

Then we would have only fields! And it would jive well with the viewmode vs editmode dichotomy, and content vs system, and public vs admin. 

Is there a particular reason why we should have the special current field that accepts multiple tags? that is the only difference from regular fields, right?

**Comment by tobibeer** [comment]
I like the way you tried explore these bits. However, you possibly have been pushing this a wee bit over the edge with aliasing.

I really was **not** suggesting to understand that fieldname itself as a tiddler, but rather some related title. Only if the alias was undefined the fieldname itself would serve as that tag. So, it wouldn't be both fieldname AND tiddler-alias, but fieldname OR tiddler-alias... to not unduly complicate things for no good reason other than "because we can".

Imho, the popup would show something more like this, one column, so far!...

* **Authors**
* **Lewis Carrol**
* author:
    * **Alice Adventures In Wonderland**
    * **Through The Looking Glass**
* tagging:
    * **Childrens Author**
    * **Englishman**
    * **Favourite Authors**

**Comment by twMat** [comment]
>I really was not suggesting to understand that fieldname itself as a tiddler, but rather 

While I did put in a disclaimer in the Note under @tobibeer for exactly this reason, I do actually wonder if it is such a bad idea. While fields *currently* are in practice typically more, hm, "systemic nature", I believ this is a direct *consequence* of their current limitations. If the proposal with tag-fields was implemented I cannot see why we would not appreciate a tiddler *author* as much as if it was a regular tag.

If we *do* implement this concept to use fields as tags, then *in practice* I expect us to see a separation between fields used for public info vs system info. With tag-fields used almost exactly like tags, why would we *not* need tiddlify fieldname as much as we need for current tags to be tiddlers?

In fact, a primary use for tags is, of course, to use them as category labels... but this is *exactly* what a field name is. 

Am I wrong? Or could it be that our innate resistence to this is because we think of fields being used as they are used today, and that there may be some possible implementation difficulties in tiddlifying field names?

**Comment by tobibeer** [comment]
> With tag-fields used almost exactly like tags, why would we not need tiddlify fieldname as much as we need for current tags to be tiddlers?

Such "tag-fields" are specifically defined as semantic lookup relation, defining a child-to-parent relationship. If any fieldname was to be "linkified", it is only these tag-fields, not all fields.

When linking a tag-field to something, it makes no sense to have **two** target tiddlers that define this type of relationship, it's either an alias **Authors** or, as a fallback, the fieldname **author**, not both.

**Comment by Jermolene** [comment]
> Then we would have only fields!

We already only have fields: a tiddler is a bundle of fields, one of which (by convention) we treat as a list of tags. There's nothing special about tags; they are just a field.

Anyhow, I'm afraid I'm really not following this discussion at all easily. I'd be grateful if you could try to explain it in functional terms. What is a tag-field, for instance?

**Comment by twMat** [comment]
Thanks for replying Jeremy! (also the earlier reply).

>We already only have fields: 

Yes, what I mean is we would only have fields that are used in the normal(?) way where the value is a single piece of information. Tags would consist of multiple instances of *name:value* pairs instead of one *name:"valuelist"*

My point "to only have fields" also refers to the users everyday perception of the matters: Currently, tags *appear* to be something other than fields because they are obviously used differently. If we instead have "viewmode fields" (which is what is referred to as "tag-fields" in this thread) and "editmode fields" these would instead probably be perceived as one thing, but with and option for it to be visible or hidden from viewmode.

>What is a tag-field, for instance?

Given the above elaborations I most likely don't understand this seemingly simple question of yours so could you please just rephrase it or explain a bit more what is unclear. Or maybe the term "tag-field" is confusing? We just mean a regular field, i.e a name:value pair, but call it "tag-field" to, for the sake of this disucssion, differentiate it from how fields and tags are currently manifested. We could call it a "viewmode field" instead. Does this answer your question?

**Comment by twMat** [comment]
@tobibeer 
>When linking a tag-field to something, it makes no sense to have two target tiddlers that define this type of relationship, it's either an alias Authors or, as a fallback, the fieldname author, not both.

Yes, sorry, I do mean that only the actual fieldname is linikified but/and that Authors is an alias. Not two target tiddlers.

**Comment by tobibeer** [comment]
> What is a tag-field?

In simple terms, a tag-field is a field that behaves like the default tag field and is qualified by its field title.

**Comment by Jermolene** [comment]
> Given the above elaborations I most likely don't understand this seemingly simple question of yours so could you please just rephrase it or explain a bit more what is unclear. Or maybe the term "tag-field" is confusing? We just mean a regular field, i.e a name:value pair, but call it "tag-field" to, for the sake of this disucssion, differentiate it from how fields and tags are currently manifested. We could call it a "viewmode field" instead. Does this answer your question?

That makes it sound like a tag-field is a different name for a field?

It also sounds like the proposal includes the idea that a tiddler could have multiple fields with the same name. That is actually something I've thought about, and agree that it would have a certain elegance to the way that we'd then implement tags. However, I think it wouldn't be possible to make this change in a backwards-compatible way.

**Comment by Jermolene** [comment]
> In simple terms, a tag-field is a field that behaves like the default tag field and is qualified by its field title.

So the value of a tag field is a list of tiddler titles? What do you mean by "qualified by its field name"?

**Comment by tobibeer** [comment]
It means that the field name **author** can be leveraged systematically to understand the contents of this field as relating to a well defined relation of a something, e.g. publication, to an **author**, something that doesn't exist with plain tags.

**Comment by Jermolene** [comment]
> It means that the fieldname author can be leveraged systematically to understand the contents of this field as relating to a well defined relationship, e.g. of Publications to Authors, something that doesn't exist with plain tags that way ...hence a "semantic" relationship.

Right, that just sounds like how we use fields to me. I'm still thinking that what this ticket is really about is exploring new ways of exposing the existing fields functionality.

**Comment by tobibeer** [comment]
> I'm still thinking that what this ticket is really about is exploring new ways of exposing the existing fields functionality.

Indeed. Here, the main proposal is to have certain fields listed as qualified tags defining tagging relationships.

**Comment by Jermolene** [comment]
>  Here, the main proposal is to have certain fields in the taglist listed as tags defining qualified tagging relationships.

The idea of classifying them as a type of tag seems to me to give rise to enormous confusion. Tags are nice and tightly defined at the moment. It also moves a long, long way from the functionality conventionally associated with tags. It complicates every single place in the code that touches tags at the moment. For example, the 'tagging' filter would need to look at these relations too. If not, what name would we use for the equivalent operator for working with the new-style tags?

Why not call them "relations" or something?

**Comment by tobibeer** [comment]
> Why not call them "relations" or something?

Yes, that would be a naming convention. We can call them relationships or simply "lookups" and they would exist, in fact, in addition to and independent of tags, so not be listed via the tagging macro or a tagging filter! That is reserved for the stanard tags field.

What would be extended are templates like the tag template for tiddlers which we wish to list such lookup-field-popups.

We wouldn't even need any new filter as we can already do the required queries...

* `[all[current]listed[author]]`
    * tiddlers for which the author field contains a reference to the current tiddler (being an author tiddler)

* `[all[current]list[author]]`
    * tiddlers listed in the author field, thus the authors associated with this tiddler

**Comment by Jermolene** [comment]
> Yes, that would be a naming convention. 

I think if we're going to introduce radically new features we really shouldn't try to shoehorn them into the same name as existing, well-understood features!

> We wouldn't even need any new filter as we can already do the required queries...

That goes back to what I was saying: all of this can be explored with custom segments for the view and edit templates.

**Comment by tobibeer** [comment]
I'll do some demo(s) and we can see where it goes from there.

**Comment by tobibeer** [comment]
Also did some renaming in the OP and called them lookup fields.

**Comment by Jermolene** [comment]
Thanks @tobibeer that's a great help.

**Comment by twMat** [comment]
>It also sounds like the proposal includes the idea that a tiddler could have multiple fields with the same name.

Hm, not as far as I can tell. Here's a reversed way to see the basic proposal here (...described from my probably simplistic user perspective):

Just like normal in edit mode we create a few fields, say:
>author : Lewis Carroll
protagonist : Alice

The only difference from normal is that these are visible in view mode and they are manifested as tagpills as per earlier illustration. Clicking on the tagpill opens a template, but I think this can be treated separately.

Now, going into the spacial case for what should *appear* like regular tags, for say "Recommended" and "Buy" one could imagine the following:
>author : Lewis Carroll
protagonist : Alice
 : Recommended
 : Buy

the blank fieldname is treated in a special way. The system adds a default fieldname, perhaps "relation" as you suggest and with some identifier, such as a mangled tiddler title, and a local counter eg:

> through_the_looking-glass.relation#1 : Recommended
 through_the_looking-glass.relation#2 : Buy

...and/but when the tagpill is displayed, this fieldname is hidden effectively making the tagpill appear like current tags do.

In addition, it is a matter to differentiate between which fields in a tiddler that should be used for this and which should behave like the current user defined fields do. Maybe just prepend with e.g "visible." or some something to the fieldname when it is created.

>I'm still thinking that what this ticket is really about is exploring new ways of exposing the existing fields functionality.

100%. It's just that if they behave like I've described here, I think they might outplay the role of current tags...

**Comment by twMat** [comment]
Yes, thank you Tobias !

**Comment by tobibeer** [comment]
>> author : Lewis Carroll
>> protagonist : Alice
>>: Recommended
>>: Buy

> the blank fieldname is treated in a special way. The system adds a default fieldname, perhaps "relation" as you suggest and with some identifier, such as a mangled tiddler title, and a local counter eg:

Not a good idea. That is what the standard tags are for!

We don't need any "special handling" of this kind. The idea of a **lookup** relation is precisely that it is qualified via the fieldname whereas the default tags define a kind of relation that is not further unqualified. There is no point in having fields with no fieldname, which sure won't work in html attributes either.

**Comment by twMat** [comment]
>Not a good idea. That is what the standard tags are for!

Part of my idea is exactly that, i.e to be able to replace standard tags with one unified solution for both tagging and user-defined fielding. But, I'll forget about it for now.

>There is no point in having fields with no fieldname,
Well, I did describe *would* be a fieldname. But, never mind :-)

BTW, also thanks to @Jermolene for the discussion on this!

**Comment by pmario** [comment]
@Jermolene .. imo should be labeled "discussion" and "feature request"
